core/graph/src/main/java/uk/gov/gchq/gaffer/graph/hook/FunctionAuthoriser.java

```diff
@@ -0,0 +1,216 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(FunctionAuthoriser.class);
+    private List<Class<? extends Function>> unauthorisedFunctions = new ArrayList<>();
+    private List<Pattern> authorisedFunctionPatterns = new ArrayList<>();
+    private List<Pattern> unauthorisedFunctionPatterns = new ArrayList<>();
+
+    public FunctionAuthoriser() {
+    }
+
+    public FunctionAuthoriser(final List<Class<? extends Function>> unauthorisedFunctions, final List<Pattern> unauthorisedFunctionPatterns, final List<Pattern> authorisedFunctionPatterns) {
+        this.setUnauthorisedFunctions(unauthorisedFunctions);
+        this.setAuthorisedFunctionPatterns(authorisedFunctionPatterns);
+        this.setUnauthorisedFunctionPatterns(unauthorisedFunctionPatterns);
+    }
+
+    @Override
+    public void preExecute(final OperationChain<?> opChain, final Context context) {
+        String chainString;
+        try {
+            chainString = new String(JSONSerialiser.serialise(opChain));
+        } catch (final SerialisationException e) {
+            // This should never happen in real life as operation chains should
+            // always be json serialisable. However this could happen if using a
+            // mock in testing. To account for this, it will be logged.
+            LOGGER.warn("Failed to serialise operation chain: " + opChain);
+            return;
+        }
+
+        if (unauthorisedFunctions != null) {
+            checkNoBlacklistedFunctionsArePresent(chainString);
+        }
+        if (authorisedFunctionPatterns != null || unauthorisedFunctionPatterns != null) {
+            checkAllFunctionsUsedAppearInCorrectPatterns(chainString);
+        }
+    }
+
+    private List<Predicate> convertPatternsToPredicates(final List<Pattern> patterns) {
+        return patterns.stream()
+                .map(Pattern::asPredicate)
+                .collect(Collectors.toList());
+    }
+
+    private void checkAllFunctionsUsedAppearInCorrectPatterns(final String chainString) {
+        boolean noWhitelistPatterns = (authorisedFunctionPatterns == null || authorisedFunctionPatterns.isEmpty());
+        boolean noBlacklistPatterns = (unauthorisedFunctionPatterns == null || unauthorisedFunctionPatterns.isEmpty());
+
+        Predicate<String> isAuthorised = createAuthorisationFunction(noBlacklistPatterns, noWhitelistPatterns);
+
+        final JsonNode node;
+        try {
+            node = JSONSerialiser.getJsonNodeFromString(chainString);
+        } catch (final SerialisationException e) {
+            // This should never happen as the string is derived from a
+            // serialised object. Therefore throw an exception here.
+            throw new RuntimeException("Failed to convert serialised operation" +
+                    " chain into a JsonNode", e);
+        }
+
+        List<String> classNames = node.findValuesAsText("class");
```

Comment:
The downside of using this mechanism, is you are assuming that all Functions will be serialised with a 'class' field. They should be, but if they are wrapped in some custom object and the JsonSubType annotation changes it, then it could be serialised differently.

Reply:
Not sure how you could do this without doing expensive serialisation